Extractor/ServerSubscriptionManager.cs

```diff
@@ -0,0 +1,105 @@
+using Microsoft.Extensions.Logging;
+
+using Opc.Ua;
+using Opc.Ua.Client;
+
+namespace Cognite.OpcUa
+{
+    public class ServerSubscriptionManager
+    {
+        private readonly ILogger<ServerSubscriptionManager> _logger;
+        private readonly UAClient _uaClient;
+        private readonly ServerNamespacesToRebrowseConfig? _config;
+
+        private readonly uint[] attributes = new[]
+        {
+            Attributes.NodeId,
+            Attributes.DisplayName,
+            Attributes.DataType,
+            Attributes.NodeClass,
+            Attributes.Description
+        };
+
+        public ServerSubscriptionManager(
+            ILogger<ServerSubscriptionManager> logger,
+            UAClient uaClient,
+            ServerNamespacesToRebrowseConfig? config
+        )
+        {
+            _logger = logger;
+            _uaClient = uaClient;
+            _config = config;
+        }
+
+        public async Task CreateCustomServerSubscriptions(UAExtractor extractor, CancellationToken token)
+        {
+            var serverNode = await _uaClient.GetServerNode(token);
+
+            List<NodeId> nodeIds = new List<NodeId>();
+            var targets = _config?.NamespaceNodes.Select(name => name.ToNodeId(_uaClient));
+
+            await _uaClient.Browser.BrowseDirectory(
+                new[] { serverNode.Id },
+                (refDef, id) =>
+                {
+                    var nodeId = (NodeId) refDef.NodeId;
+                    if (targets.Contains(nodeId))
+                    {
+                        nodeIds.Add(nodeId);
+                        _logger.LogInformation(
+                            "Subscription to a rebrowse on node {node} with {id} is now set",
+                            refDef.DisplayName.ToString(), nodeId
+                        );
+                    }
+                },
+                token,
+                maxDepth: -1,
+                doFilter: false
+            );
+
+            var readValueIds = new ReadValueIdCollection(
+                nodeIds.SelectMany(
+                    node => attributes.Select(att => new ReadValueId { AttributeId = att, NodeId = node })
+                )
+            );
+            var results = await _uaClient.ReadAttributes(readValueIds, nodeIds.Count, token, "server subscriptions");
+            var nodes = new Dictionary<NodeId, UAHistoryExtractionState>();
+
+            for (var id = 0; id < nodeIds.Count; id++)
+            {
+                var nc = (NodeClass)results[id * attributes.Count() + 3].Value;
+                if (nc != NodeClass.Variable) continue;
+                var rawDt = results[id * attributes.Count() + 2].GetValue(NodeId.Null);
+
+                nodes[nodeIds[id]] = new ServerItemSubscriptionState(_uaClient, nodeIds[id]);
+            }
+
+            await _uaClient.AddSubscriptions(nodes.Values, "NodeMetrics",
+               async (MonitoredItem item, MonitoredItemNotificationEventArgs _) => await extractor.Rebrowse(),
```

Comment:
This is extremely dangerous for multiple reasons.

First of all, this blocks the entire event queue, so no new data arrives while Rebrowse is running, which remember, can take a long time.

Secondly, if you receive 100 notifications at the same time here, which isn't actually that unlikely, you will rebrowse the OPC-UA server _100_ times, which would be bad.

Instead, do as we discussed and trigger rebrowse on the looper. This does not block, and also does not run more than once in parallel, if new notifications arrive while rebrowse is running, it will simply browse once more when it is done.

Reply:
It also does not take into consideration that you will receive notifications for the initial value of the node.

Reply:
I'll work on moving this to the looper class.

Reply:
It's better now, but we should look into preventing rebrowse being triggered from the initial value. Perhaps you can check if the data point received has a source timestamp/server timestamp greater than the start of the extractor? No point browsing for an update that we have already covered.